File: ZIMO 2/azimo2/swift/chamada.c
#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h> // Necessário para dlopen, dlsym, dlclose

// Define o tipo da função de assembly
typedef int32_t (*AssemblyFunc)(int32_t);

// Função para carregar a biblioteca e chamar a função Assembly
int32_t callAssemblyFunction(int32_t input) {
    // Determine o caminho da biblioteca dinâmica
    const char *libraryPath;
#if defined(_WIN32) || defined(_WIN64)
    libraryPath = "libassembly.dll"; // Para Windows
#else
    libraryPath = "libassembly.so"; // Para Linux/Unix
#endif

    // Carregar a biblioteca
    void *dylib = dlopen(libraryPath, RTLD_LAZY);
    if (!dylib) {
        fprintf(stderr, "Erro ao carregar a biblioteca: %s\n", dlerror());
        return -1;
    }

    // Procura a função na biblioteca
    AssemblyFunc assemblyFunction = (AssemblyFunc)dlsym(dylib, "assembly_function");
    if (!assemblyFunction) {
        fprintf(stderr, "Erro ao encontrar a função: %s\n", dlerror());
        dlclose(dylib);
        return -1;
    }

    // Chama a função Assembly com o valor desejado
    int32_t result = assemblyFunction(input);

    // Fecha a biblioteca
    dlclose(dylib);

    return result;
}

// Chamar a função com um valor de entrada
int main() {
    int32_t input = 10;
    int32_t result = callAssemblyFunction(input);
    printf("Resultado da função Assembly: %d\n", result);
    return 0;
}

File: chamada.swift
import Foundation

// Definir o tipo de função da biblioteca em Assembly
typealias AssemblyFunc = @convention(c) (Int32) -> Int32

// Função principal
func main() {
    // Carregar a biblioteca compilada em Assembly (ajuste o caminho conforme necessário)
    let dylibPath: String
    #if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
    dylibPath = "/path/to/libassembly.dylib" // Caminho para a biblioteca no macOS
    #elseif os(Linux)
    dylibPath = "/path/to/libassembly.so" // Caminho para a biblioteca no Linux
    #elseif os(Windows)
    dylibPath = "C:\\path\\to\\libassembly.dll" // Caminho para a biblioteca no Windows
    #else
    fatalError("Sistema operacional não suportado")
    #endif
    
    // Tentar carregar a biblioteca
    guard let dylib = dlopen(dylibPath, RTLD_LAZY) else {
        print("Erro ao carregar a biblioteca: \(String(cString: dlerror()))")
        return
    }
    
    // Obter a referência para a função Assembly
    let assemblyFunction: AssemblyFunc = unsafeBitCast(dlsym(dylib, "assembly_function"), to: AssemblyFunc.self)
    
    // Chamar a função Assembly com o valor desejado
    let result = assemblyFunction(10) // Exemplo de entrada
    print("Resultado da função Assembly: \(result)")
    
    // Fechar a biblioteca
    dlclose(dylib)
}

// Chamar a função principal
main()
